src/xdp/program.c

```diff
@@ -2325,6 +2678,193 @@ XdpIrpProgramClose(
+    XDP_RULE XdpRule = {0};
+    XDP_PROGRAM_OBJECT *ProgramObject;
+    ULONG RequiredMode;
+
+    TraceEnter(
+        TRACE_CORE, "AttachingProvider=%p AttachingClient=%p", AttachingProvider, AttachingClient);
+
+    if (ClientData == NULL || ClientData->size != sizeof(IfIndex) || ClientData->data == NULL) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
 
+    IfIndex = *(UINT32 *)ClientData->data;
+
+    if (IfIndex == IFI_UNSPECIFIED) {
+        Status = STATUS_NOT_SUPPORTED;
+        goto Exit;
+    }
+
+    if (ClientDispatch->version < 1 || ClientDispatch->size < 4) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    OpenParams.IfIndex = IfIndex;
+    OpenParams.HookId.Layer = XDP_HOOK_L2;
+    OpenParams.HookId.Direction = XDP_HOOK_RX;
+    OpenParams.HookId.SubLayer = XDP_HOOK_INSPECT;
+    OpenParams.Flags = XDP_CREATE_PROGRAM_FLAG_ALL_QUEUES;
+    OpenParams.RuleCount = 1;
+    OpenParams.Rules = &XdpRule;
+
+    Status = XdpRegQueryDwordValue(XDP_PARAMETERS_KEY, L"XdpEbpfMode", &RequiredMode);
+    if (NT_SUCCESS(Status)) {
+        switch (RequiredMode) {
+        case XDP_INTERFACE_MODE_GENERIC:
+            OpenParams.Flags |= XDP_CREATE_PROGRAM_FLAG_GENERIC;
+            break;
+
+        case XDP_INTERFACE_MODE_NATIVE:
+            OpenParams.Flags |= XDP_CREATE_PROGRAM_FLAG_NATIVE;
+            break;
+
+        default:
+            Status = STATUS_INVALID_PARAMETER;
+            goto Exit;
+        }
+    }
+
+    XdpRule.Match = XDP_MATCH_ALL;
+    XdpRule.Action = XDP_PROGRAM_ACTION_EBPF;
+    XdpRule.Ebpf.Target = (HANDLE)AttachingClient;
+
+    Status = XdpProgramCreate(&ProgramObject, &OpenParams, KernelMode);
+    if (!NT_SUCCESS(Status)) {
+        goto Exit;
+    }
+
+    EbpfExtensionClientSetProviderData(AttachingClient, ProgramObject);
+
+Exit:
+
+    TraceExitStatus(TRACE_CORE);
+
+    return Status;
+}
+
+static
+NTSTATUS
+EbpfProgramOnClientDetach(
+    _In_ const EBPF_EXTENSION_CLIENT *DetachingClient
+    )
+{
+    XDP_PROGRAM_OBJECT *ProgramObject = EbpfExtensionClientGetProviderData(DetachingClient);
+
+    TraceEnter(TRACE_CORE, "ProgramObject=%p", ProgramObject);
+
+    ASSERT(ProgramObject != NULL);
+
+    XdpProgramClose(ProgramObject);
+
+    TraceExitSuccess(TRACE_CORE);
     return STATUS_SUCCESS;
 }
+
+NTSTATUS
+XdpProgramStart(
+    VOID
+    )
+{
+    const EBPF_EXTENSION_PROVIDER_PARAMETERS EbpfProgramInfoProviderParameters = {
+        .ProviderModuleId = &EbpfXdpProgramInfoProviderModuleId,
+        .ProviderData = &EbpfXdpProgramInfoProviderData,
+    };
+    const EBPF_EXTENSION_PROVIDER_PARAMETERS EbpfHookProviderParameters = {
+        .ProviderModuleId = &EbpfXdpHookProviderModuleId,
+        .ProviderData = &EbpfXdpHookProviderData,
+    };
+    DWORD EbpfEnabled;
+    NTSTATUS Status;
+
+    TraceEnter(TRACE_CORE, "-");
+
+    //
+    // eBPF is disabled by default while reliability bugs are outstanding.
```

Comment:
I wonder if it's worth having an XDP bug to reference, that points to all the eBPF bugs that are blockers.

Reply:
Yeah, I filed meta bug #173 so we don't forget to enable eBPF.